### PR TITLE
Rotational energy line test

### PR DESCRIPTION
**Pull request by matthiasto**
- Adds a rotational energy line test: helps to checks eg. for numerical grid independence 
  - adds few helper functions and some input data
- Update ini file procedure for the energy line test and the rotation test

You can try it with the avaTripleBowl

### CODE REVIEW

File: avaframe/in3Utils/generateTopo.py [hunk 1]
```diff
@@ -336,26 +336,116 @@ def parabola(cfg):
...
+    # compute polar coordinates
+    r = np.sqrt(xv**2 + yv**2)
+    theta = np.arctan2(-yv, xv)
+
+    # add parabola aligned with x (going from left to right)
+    phi = math.pi
+    # rotation of the polar coord system to be aligned with the parabola direction
+    gamma = theta - phi
+    gamma = np.where(gamma < 0, gamma+2*math.pi, gamma)
+    gamma = np.where(gamma >= 2*math.pi, gamma-2*math.pi, gamma)
```
[the remark this answers is not in the document]
Comment (fso42) [reply avaframe/in3Utils/generateTopo.py:394]: These are valid concerns, this is repeated 3 times in the exact same way - make a function f(phi)

File: avaframe/in3Utils/generateTopo.py [hunk 2]
```diff
@@ -336,26 +336,116 @@ def parabola(cfg):
...
+    s = r * np.cos(gamma)
+    # shift this so that origin is at the top of the parabola
+    s = -s + fLen + fFlat
+    # apply the parabola to the corresponding part of the dem
+
+    mask = np.ones(np.shape(s))
+    mask[np.where(theta < 2*math.pi/3)] = 0
+    mask[np.where(theta <= -5*math.pi/8)] = 1
+    mask[np.where(s > fLen)] = 0
+    zv = zv + (A * s**2 + B * s + C)*mask
```
[the remark this answers is not in the document]
Comment (fso42) [reply avaframe/in3Utils/generateTopo.py:405]: same

File: avaframe/in3Utils/generateTopo.py [hunk 3]
```diff
@@ -336,26 +336,116 @@ def parabola(cfg):
...
+    s = r * np.cos(gamma)
+    # shift this so that origin is at the top of the parabola
+    s = -s + fLen + fFlat
+    # apply the parabola to the corresponding part of the dem
+
+    mask = np.ones(np.shape(s))
+    mask[np.where(theta > 2*math.pi/3)] = 0
+    mask[np.where(theta < math.pi/24)] = 0
+    mask[np.where(s > fLen)] = 0
+    zv = zv + (A * s**2 + B * s + C)*mask
```
[the remark this answers is not in the document]
Comment (fso42) [reply avaframe/in3Utils/generateTopo.py:423]: same

File: avaframe/in3Utils/generateTopo.py [hunk 4]
```diff
@@ -336,26 +336,116 @@ def parabola(cfg):
...
+
+    # add parabola aligned with x (going from left to right)
+    phi = -math.pi/4
+    # rotation of the polar coord system to be aligned with the parabola direction
+    gamma = theta - phi
+    gamma = np.where(gamma < 0, gamma+2*math.pi, gamma)
+    gamma = np.where(gamma >= 2*math.pi, gamma-2*math.pi, gamma)
```
[the remark this answers is not in the document]
Comment (fso42) [reply avaframe/in3Utils/generateTopo.py:430]: same

File: avaframe/tests/test_com1DFA.py [hunk 5]
```diff
@@ -1623,6 +1623,23 @@ def test_runCom1DFA(tmp_path, caplog):
...
+    print(simDF.index)
+    print(simDF.columns)
+    assert 'pft' in resTypeList
+    assert 'pfv' in resTypeList
+    assert 'ppr' in resTypeList
+    assert 'relKot_c9fff169d9_ent_dfa' in simDF['simName'].to_list()
+    assert 'relKot_728a70555e_null_dfa' in simDF['simName'].to_list()
+    with pytest.raises(AttributeError) as e:
+        dem, simDF, resTypeList = com1DFA.runOrLoadCom1DFA(avalancheDir, cfgMain, runDFAModule=True, cfgFile='',
+                                                           deleteOutput=False)
```
Comment (awirb) [inline avaframe/tests/test_com1DFA.py:1644]: maybe check if e.g. benchmarks/avaHockeyChannelPytest would work

File: docs/moduleAna1Tests.rst [hunk 6]
```diff
@@ -375,6 +375,8 @@ Artificial viscosity can lead to viscous dissipation leading
 to shorter runouts then what the energy solution predicts.
 Finally, the effect of the pressure force can be studied, especially the
 effect of the computation options.
+The results of this test can also be reused for other purposes or in other
+test such as in the rotation test described bellow (:ref:`Rotation test`)
```
Comment (awirb) [inline docs/moduleAna1Tests.rst:379]:
```suggestion
test such as in the rotation test described below (:ref:`Rotation test`)
```

File: docs/moduleAna1Tests.rst [hunk 7]
```diff
@@ -394,9 +396,67 @@ and the runout angle and distance errors as well as the velocity altitude error
...
+Rotation test
+==============
 
+The rotation test aims at verifying that a DFA computation module produces similar results,
+if not identical, independently of the underlying mesh or grid orientation used for the computation.
+Indeed, numerical solvers using any sort of mesh based discretization or interpolation method
+tend to give different results for the same physical problem, boundary conditions and initial
+conditions if the mesh or grid orientation is different.
+In this test, the same physical problem is fed to the DFA module changing only the grid orientation.
+The energy line test is applied to each of the simulations and the runouts compared. An AIMEC
```
Comment (awirb) [inline docs/moduleAna1Tests.rst:410]:
```suggestion
The energy line test is applied to each of the simulations and the runouts are compared. An AIMEC
```

File: docs/moduleAna1Tests.rst [hunk 8]
```diff
@@ -394,9 +396,67 @@ and the runout angle and distance errors as well as the velocity altitude error
...
+analysis of the peak results (previously rotated so that they have the same flow direction) is also
```
Comment (awirb) [inline docs/moduleAna1Tests.rst:411]:
```suggestion
analysis of the peak results (previously rotated so that the results are aligned) is also
```

File: docs/moduleAna1Tests.rst [hunk 9]
```diff
@@ -394,9 +396,67 @@ and the runout angle and distance errors as well as the velocity altitude error
...
+carried out to give a more detailed idea of the spatial difference between the simulations.
```
Comment (awirb) [inline docs/moduleAna1Tests.rst:412]:
```suggestion
carried out to give a more detailed idea of the spatial differences between the simulations.
```

File: docs/moduleAna1Tests.rst [hunk 10]
```diff
@@ -394,9 +396,67 @@ and the runout angle and distance errors as well as the velocity altitude error
...
+at the same distance from the origin and a ring shaped entrainment feature
+(as shown on :numref:`fig-rotation-test-inputs`).
+
+.. _fig-rotation-test-inputs:
+
+.. figure:: _static/rotationTestOutputs1.png
+    :width: 90%
+
+    Example of input data used for the rotation test and peak flow thickness result produced by the com1DFA module.
+    In the background one sees the dem of a bowl, in dashed areas (small circles) the three release features
```
Comment (awirb) [inline docs/moduleAna1Tests.rst:446]:
```suggestion
    In the background the bowl DEM is shown, in dashed areas (small circles) the three release features
```

File: docs/moduleAna1Tests.rst [hunk 11]
```diff
@@ -394,9 +396,67 @@ and the runout angle and distance errors as well as the velocity altitude error
...
+    and in blue (big circle) the entrainment feature. One release feature is aligned with the x axis (rel0),
+    the other is rotated 120° clockwise (rel120) and the last on 225° clockwise (rel225). All three input
+    scenarios are identical.
```
Comment (awirb) [inline docs/moduleAna1Tests.rst:449]: identical in terms of size  or maybe better release mass?

File: docs/moduleAna1Tests.rst [hunk 12]
```diff
@@ -394,9 +396,67 @@ and the runout angle and distance errors as well as the velocity altitude error
...
+
+One then needs to specify the the com1DFA configuration (through the ``com1DFACfg.ini`` and its local version)
+as well as the AIMEC configuration (through the ``ana3AIMECCfg.ini`` and its local version) and
+the path generation configuration (through the ``pathGenerationCfg.ini`` and its local version).
+
+To run
+-------
+
+A workflow example is given in :py:mod:`runScripts.runRotationTest.py`.
+Some example input data is given in avaTripleBowl.
```
Comment (awirb) [inline docs/moduleAna1Tests.rst:459]: maybe add what kind of geometry this is?

File: avaframe/ana1Tests/energyLineTest.py [hunk 13]
```diff
@@ -209,18 +198,19 @@ def generateCom1DFAEnergyPlot(avalancheDir, energyLineTestCfg, avaProfileMass, d
     ax3.set_yticks([avaProfileMass['z'][-1]-z0, zIntersection-z0])
     ax3.xaxis.set_major_formatter(FormatStrFormatter('%.2f'))
     ax3.yaxis.set_major_formatter(FormatStrFormatter('%.2f'))
-    text = ('Runout s diff : %.2f m \nRunout z diff : %.2f m \nRunout angle diff : %.4f° \nvelocity height rmse : %.2f m \n(energy line - alpha line)' %
+    text = ('Runout s diff : %.8f m \nRunout z diff : %.8f m \nRunout angle diff : %.8f° \nvelocity height rmse : %.8f m \n(energy line - alpha line)' %
             (runOutSError, runOutZError, runOutAngleError, rmseVelocityElevation))
     text_box = AnchoredText(text, frameon=False, loc=1, pad=0.5,
                             prop=dict(fontsize=pU.fs))
+    log.info(text)
```
Comment (fso42) [inline avaframe/ana1Tests/energyLineTest.py:205]: I think this should be debug level

File: avaframe/ana1Tests/energyLineTest_com1DFACfg.ini [hunk 14]
```diff
@@ -32,11 +32,11 @@ sphKernelRadiusTimeStepping = True
 # courant number
 cMax = 0.02
 # stopCriterion (based on massFlowing or kinEnergy)
-stopCritType = massFlowing
+stopCritType = kinEnergy
 # if based on massFlowing, specify the velocity threshold for flowing mass (m/s)
 uFlowingThreshold = 0.0
 # stopCriterion (stops when massFlowing<0.01*peakMassFlowing or ke<0.01*pke)
-stopCrit = 0.0
+stopCrit = 0.00
```
Comment (fso42) [inline avaframe/ana1Tests/energyLineTest_com1DFACfg.ini:39]: unnesessary 0

File: avaframe/ana1Tests/rotationTest.py [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+        row index of the reference simulation in the simDF
+
+    Returns
+    -------
+    simDF : dataFrame
+        DFA simulation dataFrame updated with the energy line test results
+    flagMass : boolean
+        should the aimec mass analysis be done (switched to true if it is an entrainment simulation)
+    """
+    # read energy line config. ToDo should we only read the DFAPath config and remove the energy line one?
```
Comment (fso42) [inline avaframe/ana1Tests/rotationTest.py:57]: Should this TODO be addressed right away?

File: avaframe/ana1Tests/rotationTest.py [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+    energyLineTestCfgFile = pathlib.Path('ana1Tests', 'energyLineTest_com1DFACfg.ini')
+    energyLineTestCfg, modInfo = cfgUtils.getModuleConfig(com1DFA, fileOverride=energyLineTestCfgFile, modInfo=True)
+    inDir = avalancheDir / 'Outputs' / 'com1DFA'
+    outDir = avalancheDir / 'Outputs' / 'com1DFARotated'
+    fU.makeADir(outDir)
+    # get reference angle
+    simName = simDF.loc[refSimRowHash, 'simName']
+    relName = (simName.split('_'))[0]
+    thetaRef = float(relName[3:])
+    for simHash in simDF.index:
```
Comment (fso42) [inline avaframe/ana1Tests/rotationTest.py:67]: I think this (and other occurrences) should be actually named rowSimHash to avoid confusion with the original simHash

File: avaframe/in3Utils/generateTopoCfg.ini [hunk 17]
```diff
@@ -27,6 +28,9 @@ demType = HS
 # distance to point where slope transitions into flat plane [m] - required for PF, HX if meanAlpha is not provided
 fLens = 2250
 
+# legnth of the flat foreground for the triple parabola for TPF
```
Comment (fso42) [inline avaframe/in3Utils/generateTopoCfg.ini:31]:
```suggestion
# length [m] of the flat runout for the triple parabola for TPF
```

File: avaframe/in3Utils/geoTrans.py [hunk 18]
```diff
@@ -900,6 +896,49 @@ def cartToSpherical(X, Y, Z):
...
+        --------
+        rotatedRaster: dict
+            rotated raster dictionary
+    """
+
+    # convert to rad if provided as degree
+    if deg:
+        theta = np.radians(theta)
+
+    # create raster gird with origin 0,0
```
Comment (fso42) [inline avaframe/in3Utils/geoTrans.py:922]:
```suggestion
    # create raster grid with origin 0,0
```

File: avaframe/tests/test_generateCompareReport.py [hunk 19]
```diff
@@ -56,14 +56,14 @@ def test_copyAimecPlots(tmp_path):
...
-    plotFiles = [{'aimec plot 1': pathPlot}, {'aimec plot 2': pathPlot}]
+    plotFiles = [{'aimec plot 1': {'sim1': pathPlot}}, {'aimec plot 2': {'sim2': pathPlot}}]
     testName = 'avaTest'
     plotPaths = {}
 
     # call function to be tested
     plotPaths = gR.copyAimecPlots(plotFiles, testName, outDir, plotPaths)
     testFile = outDir / 'avaTest__release1HX_ent_dfa_3f8d2d9327_pft.png'
-
+    print(plotPaths)
```
Comment (fso42) [inline avaframe/tests/test_generateCompareReport.py:66]: remove this

File: avaframe/tests/test_generateReport.py [hunk 20]
```diff
@@ -59,21 +59,23 @@ def test_writeReport(tmp_path):
...
-        lineVals.append(line)
+    reportFile1 = pathlib.Path(tmp_path, 'fullSimulationReport.md')
+    with open(reportFile1, 'r') as pfile:
+        lines = pfile.readlines()
+        lineVals = []
+        for line in lines:
+            lineVals.append(line)
 
-    reportFile.close()
+    print(lineVals)
```
Comment (fso42) [inline avaframe/tests/test_generateReport.py:78]: This is for debugging?

File: avaframe/tests/test_generateReport.py [hunk 21]
```diff
@@ -91,25 +93,25 @@ def test_writeReport(tmp_path):
...
-    assert lineVals2[24] == '| ----------| \n'
-    assert lineVals2[-3] == '![pft](pft.png) \n'
-    assert lineVals2[37] == '![Peak Pressure Field of my test](release1HS2_entres_dfa_0.750_pft.png) \n'
+    reportFile2 = pathlib.Path(tmp_path, 'simulationNumber2.md')
+    with open(reportFile2, 'r') as pfile:
+        lines = pfile.readlines()
+        lineVals2 = []
+        for line in lines:
+            lineVals2.append(line)
+        print('lineVals2', lineVals2)
```
Comment (fso42) [inline avaframe/tests/test_generateReport.py:107]: same here?

File: avaframe/tests/test_geoTrans.py [hunk 22]
```diff
@@ -576,3 +577,17 @@ def test_rotate():
...
+
+def test_rotateRaster():
+    rasterData = np.zeros((5, 5))
+    rasterData[2, 0:2] = 1
+    rasterData[2, 3:4] = -1
+    header = {'xllcenter': -2, 'yllcenter': -2, 'cellsize': 1, 'ncols': 5, 'nrows': 5}
+    rasterDict = {'header': header, 'rasterData': rasterData}
+    theta = 90
+    print(header)
+    print(rasterData)
```
Comment (fso42) [inline avaframe/tests/test_geoTrans.py:590]: same here?

**Review by fso42 — CHANGES_REQUESTED** [review]
- [ ] Please address the long line pep's
- [ ] See if adding Kot files can be avoided by using existing files from `data`
- [ ] TripleBowl input is now given for each angle: would it be possible to automate this and avoid these files?
- [x] I think we should mark the runCompareXXXvsCom1DFAOrig as obsolete and remove them...

File: avaframe/in3Utils/generateTopo.py [hunk 23]
```diff
@@ -336,26 +336,116 @@ def parabola(cfg):
...
+    mask[np.where(theta <= -5*math.pi/8)] = 1
+    mask[np.where(s > fLen)] = 0
+    zv = zv + (A * s**2 + B * s + C)*mask
+
+    # add parabola sloped 60° with x
+    phi = math.pi/3
+    # rotation of the polar coord system to be aligned with the parabola direction
+    gamma = theta - phi
+    gamma = np.where(gamma < 0, gamma+2*math.pi, gamma)
+    gamma = np.where(gamma >= 2*math.pi, gamma-2*math.pi, gamma)
```
[the remark this answers is not in the document]
Comment (fso42) [reply avaframe/in3Utils/generateTopo.py:412]: same

File: avaframe/ana1Tests/rotationTest.py [hunk 24]
```diff
@@ -0,0 +1,240 @@
...
+                          'Rotation test AIMEC result table': {
+                          'type': 'pandasDF', 'column names': {'simName': 'Simulation name', 'relAngle': 'Angle °',
+                                                               'sRunout': 'Runout [m]', 'relMass': 'release mass [kg]',
+                                                               'finalMass': 'final mass [kg]',
+                                                               'entMass': 'entrained mass [kg]',
+                                                               'TP': 'True positive area [-]',
+                                                               'FP': 'False positive area [-]',
+                                                               'FN': 'False negative area [-]'}},
+                          'Rotation test AIMEC figures': {'type': 'image'}}
+    resTypeList = list(set(resTypeList).intersection(['ppr', 'pfv', 'pft']))
```
[the remark this answers is not in the document]
Comment (matthiasto) [reply avaframe/ana1Tests/rotationTest.py:181]: Because AIMEC can only analyze these 3. So I want to get rif of any other restype. And I am not sure all these 3 restypes are in the simDF so I remove them if not

File: avaframe/com1DFA/com1DFACfg.ini [hunk 25]
```diff
@@ -109,7 +109,7 @@ tEnd = 400
 # to use a variable time step (time step depends on kernel radius)
 sphKernelRadiusTimeStepping = False
 # Upper time step limit coefficient if option sphKernelRadiusTimeStepping is chosen.
-cMax = 0.5
+cMax = 0.01
```
[the remark this answers is not in the document]
Comment (matthiasto) [reply avaframe/com1DFA/com1DFACfg.ini:112]: You are right, it should be 0.02
because if you set sphKernelRadiusTimeStepping to True, the dt = cMax * csz = 0.02 * 5 = 0.1s

File: avaframe/tests/test_com1DFA.py (hunk 5 above)
Comment (matthiasto) [reply avaframe/tests/test_com1DFA.py:1644]: Why would this work? The avaHockeyChannelPytest does not provide the allConfig.csv

File: avaframe/log2Report/generateCompareReport.py [hunk 26]
```diff
@@ -38,11 +38,14 @@ def copyAimecPlots(plotFiles, testName, outDir, plotPaths, rel=''):
...
-            log.debug('Copied: %s to %s' % (pDict[key], plotName))
-            plotPaths[key] = plotName
+            pDict2 = pDict[key]
+            for key2 in pDict2:
+                if isinstance(pDict2[key2], pathlib.PurePath):
+                    name = pathlib.Path(pDict2[key2]).name
+                    plotName = outDir / ('%s_%s_%s' % (testName, rel, name))
+                    shutil.copy2(pDict2[key2], plotName)
+                    log.info('Copied: %s to %s' % (pDict2[key2], plotName))
+                    plotPaths[key] = plotName
```
[the remark this answers is not in the document]
Comment (matthiasto) [reply avaframe/log2Report/generateCompareReport.py:48]: seems to yes

File: avaframe/runScripts/runAna3AIMEC.py [hunk 27]
```diff
@@ -56,8 +39,26 @@ def runAna3AIMEC(avalancheDir=''):
...
     rasterTransfo, resAnalysisDF, plotDict = ana3AIMEC.mainAIMEC(pathDict, inputsDF, cfg)
-    return pathDict, rasterTransfo, resAnalysisDF
+    return pathDict, rasterTransfo, resAnalysisDF, plotDict
 
 
 if __name__ == '__main__':
-    runAna3AIMEC()
+    # Load avalanche directory from general configuration file
+    cfgMain = cfgUtils.getGeneralConfig()
+    avalancheDir = cfgMain['MAIN']['avalancheDir']
```
[the remark this answers is not in the document]
Comment (matthiasto) [reply avaframe/runScripts/runAna3AIMEC.py:48]: Then you have to call runAna3Aimec(yourAvaDir)

File: avaframe/ana1Tests/energyLineTest.py (hunk 13 above)
Comment (matthiasto) [reply avaframe/ana1Tests/energyLineTest.py:205]: Isn't it nice to also see the results in the terminal?

File: avaframe/tests/test_generateCompareReport.py (hunk 19 above)
Comment (matthiasto) [reply avaframe/tests/test_generateCompareReport.py:66]: this is the tests, we always had some... if I remove this one I have to remove all

File: avaframe/ana1Tests/rotationTest.py [hunk 28]
```diff
@@ -0,0 +1,240 @@
...
+    """Rotate the DFA results
+
+    Parameters
+    ----------
+    avalancheDir : pathlib path
+        path to avalanche directory
+    simDF : dataFrame
+        DFA simulation dataFrame
+    simHash : str
+        row index of the simulation to analyze
```
[the remark this answers is not in the document]
Comment (matthiasto) [reply avaframe/ana1Tests/rotationTest.py:105]: No it is the row hash, because fU.makeSimFromResDF now changes the index (it is so that aimec always gets a different row hash even for 2 identical simulations living in different places)

File: avaframe/com1DFA/com1DFACfg.ini [hunk 29]
```diff
@@ -109,7 +109,7 @@ tEnd = 400
 # to use a variable time step (time step depends on kernel radius)
 sphKernelRadiusTimeStepping = False
 # Upper time step limit coefficient if option sphKernelRadiusTimeStepping is chosen.
-cMax = 0.5
+cMax = 0.02
```
Comment (fso42) [inline avaframe/com1DFA/com1DFACfg.ini:112]: Is this on purpose?

File: avaframe/log2Report/generateReport.py [hunk 30]
```diff
@@ -4,27 +4,38 @@
...
 # create local logger
 # change log level in calling module to DEBUG to see log messages
 log = logging.getLogger(__name__)
 
 
 def copyPlots2ReportDir(reportDir, plotDict):
-    """ copy the plots to report directory """
+    """ copy the plots to report directory
+    The plots are in a dictionary:
+    plotDict = {'plot1': PureṔath to plot1, 'plot2': PureṔath to plot2...}
```
Comment (fso42) [inline avaframe/log2Report/generateReport.py:21]:
```suggestion
    plotDict = {'plot1': PurePath to plot1, 'plot2': PurePath to plot2...}
```

**Review by fso42 — CHANGES_REQUESTED** [review]

File: avaframe/com1DFA/com1DFACfg.ini (hunk 29 above)
Comment (matthiasto) [reply avaframe/com1DFA/com1DFACfg.ini:112]: Yes, 0.02 means a time step of 0.1s if the mesh size is 5m which is the default value we want

**Review by fso42 — APPROVED** [review]
